Notify user when messages in unsubscribed streams are not marked as unread #23470

sanchi-t opened the pull request:
<!-- Describe your pull request here.-->
Displays a notice to inform users when the "Mark as unread from here" action is unable to mark some messages as unread because they are in unsubscribed streams.

This PR attempts to complete the original work done by https://github.com/zulip/zulip/pull/26634 completion candidate PR.

Fixes: https://github.com/zulip/zulip/issues/23470

Discussion: [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/not.20marked.20as.20unread.20banner/near/1627530)
[CZO design discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/update.20has_message_access.3F/near/1638961)
[CZO api design](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Not.20marked.20unread.20banner/near/1635508)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
Dark Mode:
![{B14FCF5B-295A-42D8-8C09-0522498DBB24}](https://github.com/user-attachments/assets/ff4e4796-c03d-47da-bb4b-ab8341572703)![{7ABBBC6C-0C5E-43A8-AB47-B97EF1824670}](https://github.com/user-attachments/assets/06d8d981-444a-411b-809d-5ffceeb560bd)

Light Mode:
![{9B830FD2-AF98-47CA-B676-90C70B94AD4B}](https://github.com/user-attachments/assets/2a97c0a1-a13e-4c44-a365-bfd49fe5ed91)
![{C0C3C585-E0F4-4158-83ED-817D709EFE7E}](https://github.com/user-attachments/assets/bd6b9bd1-392a-4f9d-9eea-f7aa05d441cc)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sanchi-t commented:
@zulipbot add label "mentor review".

sanchi-t commented:
@zulipbot label "mentor review".

alya commented:
@sanchi-t Are you still expecting mento review here, or should we proceed to maintainer review?

ryanreh99 commented on zerver/actions/message_flags.py:
```diff
@@ -283,12 +284,24 @@ def do_update_message_flags(
+                {
+                    message[1]
+                    for message in messages_in_unsubscribed_streams
+                    if message[0] in messages
```
wouldn't this condition always satisfy ?
as `messages_in_unsubscribed_streams` is from the query which filters the Messages table for the same ids.

sanchi-t replied:
Yeah you are right updated it.

ryanreh99 commented on zerver/openapi/zulip.yaml:
```diff
@@ -7134,8 +7134,26 @@ paths:
                           type: integer
                         description: |
                           An array with the IDs of the modified messages.
+                      skipped_marking_unread_stream_ids:
```
perhaps rename it to `unread_streams` ?
in future we might do other things as well, not just mark skipped.

Better to discuss this naming once in CZO

sanchi-t replied:
Apparently there was a [CZO discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Not.20marked.20unread.20banner/near/1637370)where the name was decided. Updated the code.

ryanreh99 reviewed (commented):
I left 2 comments,
Also, as seen in the screenshots you've shared. For a single stream there's an extra space before the comma but it is not there when multiple streams are present.

sanchi-t commented:
> Also, as seen in the screenshots you've shared. For a single stream there's an extra space before the comma but it is not there when multiple streams are present.

Updated the PR and screenshots.

alya commented:
We've renamed "streams" -> "channels" some time ago now. Please update the code and screenshots accordingly.

alya commented:
The string is not correct for 1 channel.

alya commented:
Also, it would be best if the channel symbol was always on the same line as the channel name.

Are we using privacy markers (globe, lock)? I didn't see any in the screenshot, so wasn't sure.

sanchi-t commented:
> The string is not correct for 1 channel. See issue description.

Yes I updated the PR.


> Also, it would be best if the channel symbol was always on the same line as the channel name.

Updated the PR to display channel symbol on the same line as channel name.


> Are we using privacy markers (globe, lock)? I didn't see any in the screenshot, so wasn't sure.

Yes, updated the screenshots to include different privacy markers.

alya commented:
Cool, the screenshots look good to me now. You can tag @amanagr for review once conflicts are resolved. (I haven't tested.)

sanchi-t commented:
@amanagr Ready for review from you.

amanagr commented on web/src/feedback_widget.ts:
```diff
@@ -157,17 +157,16 @@ export function show(opts: FeedbackWidgetOptions): void {
 
     meta.$container = $("#feedback_container");
 
```
Can you explain the need for this commit?

sanchi-t replied:
I was working locally on this PR before the commit d5eb99724294c3cdfae1a2860f75f03f8ae54fd1 was merged. I made these changes prior to rebasing.

Now these changes are not required.

amanagr commented:
LGTM other than the comment above.

amanagr commented:
Hey, @sanchi-t can you please rebase this?

amanagr commented:
Looks good to me. Tested manually.

timabbott commented on web/templates/skipped_marking_unread.hbs:
```diff
@@ -0,0 +1,23 @@
+                    {{~#if @last~}}
+                        <span> and </span>
+                    {{/if~}}
+                {{/unless~}}
```
This isn't properly internationalized; it should be possible to use `format_array_as_list` in the caller to format the list of channels properly.

sanchi-t replied:
Updated the PR.

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -336,6 +374,9 @@ function do_mark_unread_by_narrow(
                 finish_loading(messages_marked_unread_till_now);
             }
+            if (ignored_because_not_subscribed.length > 0) {
+                handle_skipped_unsubscribed_streams(ignored_because_not_subscribed);
```
How does this work when this code path is called recursively (see the FOLLOWUP_BATCH_SIZE logic)? 

Probably one wants to accumulate a set of channels by merging what was returned with the set we had previously, passing into the recursive calls, and only do this call from inside the block that has `finish_loading`.

sanchi-t replied:
You're right it was incorrectly handling the recursion. Fixed it.

timabbott commented:
Posted a couple comments on the UI code. The backend can be merged as soon as we get approval from https://chat.zulip.org/#narrow/channel/378-api-design/topic/unreads.20in.20unsubscribed.20streams/near/2009252; feel free to DM me if you see consensus there, so I can merge to avoid more renumberings.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
+  [`POST /messages/flags`](/api/update-message-flags):
+  Added `ignored_because_not_subscribed_channels` field in the response, which
+  is a list of the channels whose messages were skipped to mark as unread
+  because the user is not subscribed to them.
```
The API change looks good; opened #33617 to merge just that commit. I'll post some comments on the frontend change.

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -360,6 +411,10 @@ function do_mark_unread_by_narrow(
+                if (unsubscribed_ignored_channels.length > 0) {
+                    handle_skipped_unsubscribed_streams(unsubscribed_ignored_channels);
+                    unsubscribed_ignored_channels = [];
+                }
```
Why does this only need to happen if `loading_indicator_displayed`? 

If it does, I expect it can be refactored into being part of `finish_loading` ... but I don't see why we can assume that a single-batch call would not end up here.

timabbott commented:
Merged, thanks @sanchi-t!
